docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+
+####### Example of adoption via an infrastructure provider
+
+####### Example of an heterogeneous infrastructure, i.e. VMC on AWS backed by NLB
+
+#### TODO: Changes needed to clusterctl and existing infrastructure providers
+Notes:
+- in order to determine UX changes it is required to define packaging/installation sequence for the Load Balancer/InfraLoadBalancer (bundled with CAPI/infra providers, stand alone, both)
+- we should define also how credential for the InfraLoadBalancer should be provided
+- we should brainstorm both on the clusterctl workflow and on the github workflow (without clusterctl)
+- the “quick start” workflow should be as simple as possible  
+
+### Security Model
+
+Roles
+For the purposes of this security model, 3 common roles have been identified:
+Infrastructure provider: The infrastructure provider (infra) is responsible for the overall environment that the cluster(s) are operating in or the PaaS provider in a company.
+Management cluster operator: The cluster operator (ops) is responsible for administration of the Cluster API management cluster. They manage policies, network access, application permissions.
+Workload cluster operator: The workload cluster operator (dev) is responsible for management of the cluster relevant to their particular applications .
+
+#### Write Permissions
+
+| | Cluster | Load Balancer | AWSELBLoadBalancer | AWS IAM API | Relevant AWS APIs |
+|---|---|---|---|---|---|
+| Load Balancer Provider | No | No | Yes | No | Yes |
+| Load Balancer Controller | No | Yes | Yes | No | No |
+| Cluster Controller | Yes | Yes | No | No | No |
+| Management Cluster Operators | Yes | Yes | Yes | Yes | No |
+| Workload Cluster Operator | Yes | Yes | Yes | No | No |
+
+TODO: Credential management changes for infra providers
+
+#### Credential Management
+
+Infra load balancers need a specific credential workflow:
+Look at the infra load balancer resource for spec.CredentialRef:
+If it’s set: use it for load balancing provisioning
+If it’s not set: the infra load balancer provider tries to read credentials from the infrastructure cluster owning it:
+If the infra cluster is from the same type, use its credentialRef
+If the infra cluster is not from the same type, fallback to the credentials of infrastructure load balancer controller
+
+### Risks and Mitigations
+
+- What are the risks of this proposal and how do we mitigate? Think broadly.
+- How will UX be reviewed and by whom?
+- How will security be reviewed and by whom?
+- Consider including folks that also work outside the SIG or subproject.
+
+## Alternatives
+
+### Using Services API
+
+Kubernetes already provides a load balancer construct of Service Type=Loadbalancer. The service API allows pointing at arbitrary IPs, which could in fact be the machines of a workload cluster’s control plane.
+
+While we believe this approach has longer term benefits, there are a few reasons that it does not fit the short term requirements:
```

Comment:
> I think a major difference is in which clusters the resources, specifying the loadbalancers, are deployed.

Yes, I think this is the important distinction that I need to clarify better in the proposal.

> 
> In our case (our managed Kubernetes-as-a-Service) we currently have a clean separation:
> 
> * users of the workload cluster can define services of type LoadBalancer in the workload cluster (reconciled via ccm) (CSI also works the same way)
>   
>   * this is also somewhat relevant to be able to completely backup/restore a workload cluster
> * management cluster operators can define control plane LoadBalancer in the management cluster
> 
> If now the non-control-plane loadbalancers should be configured in the management cluster _and_ we want to enable workload clusters to configure them, we have to give workload cluster users access to the management cluster. But maybe that's just something we have to figure out downstream because it's specific to our use case / service.

I don't think we want to take over that use case, my thought around non-controlPlane load balancers was infrastructure that would have chicken/egg scenarios with the cluster creation. Say I'm standing up n clusters, and for each one I have requirements to configure both load balancing for the control plane, but also provider-specific ingress for the cluster worker nodes (or a subset of the worker nodes). While there may be some combinations of ingress solutions that could work with a subset of CCM providers to provide that functionality, that functionality isn't universal.

> 
> I think overall I really like the idea that the lb providers can be used for control-plane and non-control-plane loadbalancers. I assume the currently available cloud providers don't really fit in the big picture of CAPI-managed clusters. I'm just not entirely sure if we have clear picture yet how non-control-plane loadbalancers fit.

Hopefully the above helps clarify things, the intention is not to try and assume control over things that can be filled by CCM providers, but rather to supplement functionality where they don't currently meet it.

On a longer term, it would be great if CCM implementations were able to be run in a multi-tenancy based deployment on a management cluster and could fill the needs of both management cluster managed load balancers and workload cluster managed load balancers, however I think it would take significant effort (and time) to drive the functionality needed to support both of those deployment scenarios and use cases on the CCM side.